File: lcs.js
function LCS(s, t, m, n) {
	if (m >= s.length)
		return 0;
	if (n >= t.length)
		return 0;
	var key = m + '-' + n;
	if (cache[key])
		return cache[key];
	var value0 = 0;
	if (s.charAt(m) === t.charAt(n))
		value0 = LCS(s, t, m+1, n+1) + 1;
	var value1 = LCS(s, t, m+1, n);
	var value2 = LCS(s, t, m, n+1);
	return cache[key] = Math.max(value0, Math.max(value1, value2));
}

function solve(s, t) {
	var key = s.length + '-' + t.length + '-' + s + "-" + t;
	if (cache[key]) return cache[key];
	if (s == "" || t == "") return cache[key] = 0;
	if (s.charAt(0) == t.charAt(0))
		return cache[key] = 1 + solve(s.substring(1), t.substring(1));
	var value1 = solve(s.substring(1), t);
	var value2 = solve(s, t.substring(1));
	return cache[key] = Math.max(value1, value2);
}

var s = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA";
var t = "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB";
s = s + 'XAY' + s;
t = t + 'XBY' + t;
var cache = [];
console.log(solve(s, t));
// console.log(LCS(s, t, 0, 0));
/*
var s = "ATGCAGAGGGTACATTTGATAGCT";
var t = "TGAGCAGTGCAGGCAGGCAGCAGG";
var cache = [];
console.log(LCS("ATGCAGAGGGTACATTTGATAGCT", "TGAGCAGTGCAGGCAGGCAGCAGG", 0, 0));
var cache = [];
console.log(LCS(s, s, 0, 0));
var cache = [];
console.log(LCS(s, s+s, 0, 0));
var cache = [];
console.log(LCS(t+s, s+t, 0, 0));
var cache = [];
console.log(LCS(s+t+t+t+t+t+t+t+t+t+t+t, t+s+t+t+t+t+t+t+t+t+t+t, 0, 0));
*/

File: lcs.java
import java.util.*;

class lcs {

	static int solve(String s, String t) {
		String key = s + "-" + t;
		if (cache.containsKey(key)) return cache.get(key);
		int result = 0;
		if (s.length() == 0 || t.length() == 0) {
			result = 0;
		} else if (s.charAt(0) == t.charAt(0)) {
			result = 1 + solve(s.substring(1), t.substring(1));
		} else {
			int value1 = solve(s, t.substring(1));
			int value2 = solve(s.substring(1), t);
			result = Math.max(value1, value2);
		}
		cache.put(key, result);
		return result;
	}

	static HashMap<String,Integer> cache = new HashMap<>();

	public static void main(String [] a) {
		String s = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA";
		String t = "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB";
		System.out.println(solve(s + "X" + s, t + "X" + t));
	}
}
